Native Auth Backward Compatibility feature, Fixes AB#3287604

Yuki-YuXin opened the pull request:
Fixes [AB#{3287604}](https://dev.azure.com/IdentityDivision/Engineering/_workitems/edit/3287604)
Related common PR: https://github.com/AzureAD/microsoft-authentication-library-common-for-android/pull/2669

Native Auth Backward Compatibility feature includes:
- [Add capabilities parameter to native auth public client application initialization, Fixes AB#3287412](https://github.com/AzureAD/microsoft-authentication-library-for-android/pull/2305/commits/6d553466cab96eb9642f5dd65e63d3492b95d8b3)
- [Handle redirect across all endpoints and return the redirect reason to the developers, Fixes AB#3288191](https://github.com/AzureAD/microsoft-authentication-library-for-android/pull/2305/commits/a121610861b63f5b7b7fe3c7a5240d2c0ee4c0ea)
-  [Use parameter instead of config when initialization](https://github.com/AzureAD/microsoft-authentication-library-for-android/pull/2313)

The main addition to this PR is the testing section

Yuki-YuXin committed:
Add capabilities parameter to native auth public client application initialization, Fixes AB#3287412 (#2303)

Fixes
[AB#3287412](https://identitydivision.visualstudio.com/Engineering/_workitems/edit/3287412)

Related common PR:
https://github.com/AzureAD/microsoft-authentication-library-common-for-android/pull/2656

The Capabilities was added in the common repo however the MSAL repository is pointing to the dev branch by default, not our feature branch. Will change the MSAL feature branch to point to the Common feature branch to ensure pipeline validation of the merge to dev, but doing similar on individual branches tends to introduce unwanted reference errors

Yuki-YuXin committed:
Handle redirect across all endpoints and return the redirect reason to the developers, Fixes AB#3288191 (#2304)

Fixes
[AB#3288191](https://identitydivision.visualstudio.com/Engineering/_workitems/edit/3288191)

Related common PR:
https://github.com/AzureAD/microsoft-authentication-library-common-for-android/pull/2660

Pipeline failed: common reference

Yuki-YuXin committed:
Update submodule to point to branch feature

Yuki-YuXin committed:
update submodule reference

Yuki-YuXin committed:
Point to the feature/backward-compatibility

Yuki-YuXin committed:
Update NativeAuthPublicClientApplicationAbstractTest.kt and consuming code to include defaultCapabilities

Yuki-YuXin committed:
Append E2E test for MFA and JIT

Yuki-YuXin committed:
Fix authorities default null issue and Use new createNativeAuthPublicClientApplication

Yuki-YuXin committed:
Fix defaultCapabilities = listOf("mfa_required", "registration_required") format

Yuki-YuXin committed:
After changes under getAuthorities inside PublicClientApplicationConfiguration, mAuthorities could be null or mAuthorities.size() == 0
Make changes under getDefaultAuthority() to pass the test

Yuki-YuXin commented:
A hint for the change log update and sub module reference after common PR get merged.

Yuki-YuXin commented on msal/src/main/java/com/microsoft/identity/client/PublicClientApplicationConfiguration.java:
```diff
@@ -247,6 +248,9 @@ public void setClientId(final String clientId) {
+        if (mAuthorities == null) {
+            mAuthorities = new ArrayList<>();
+        }
         return mAuthorities;
```
May need Redmon team for confirmation since it's an important class. Or maybe there's another, better way to fix it.

**_Why the changes?_**
1. developer's authorities config cannot be null. 
2. the default status of authorities is null.
3. `getAuthorities()` throws error when authorities is null.
4. The transformation of null to empty list depends on the `defaultMsalConfiguration(R.raw.msal_default_config)` and merging, clearing operations.

_**What's the problem?**_
```kotlin
val nativeAuthConfig = NativeAuthPublicClientApplicationConfiguration()
            nativeAuthConfig.clientId = config.clientId
            val authorityObject = Authority.getAuthorityFromAuthorityUrl(
                config.authorityUrl,
                config.clientId
            )
            authorityObject.setDefault(true)
            nativeAuthConfig.getAuthorities().add(authorityObject)
            nativeAuthConfig.setChallengeTypes(challengeTypes)
            nativeAuthConfig.setCapabilities(capabilities)
```
`nativeAuthConfig.getAuthorities()` will have null pointer error if authorities is null. However, `Authorities` does not open the set method. 

Previous deprecated method
```java
return createNativeAuthApplication(
    Companion.initializeNativeAuthConfiguration(context),
    clientId,
    authority,
    redirectUri,
    challengeTypes
);
```
which use 
```java
fun initializeNativeAuthConfiguration(context: Context): NativeAuthPublicClientApplicationConfiguration {
    return initializeNativeAuthConfigurationInternal(context, null)
}
```
Since the `developerConfig` parameter is null, thus will skip the `developerConfig.validateConfiguration()` step inside `initializeNativeAuthConfigurationInternal(context: Context, developerConfig: NativeAuthPublicClientApplicationConfiguration?)` method, where `if (authorities == null || authorities.size == 0)` will throw an exception.

The authorities keep null until merging and clearing
```java
private fun initializeNativeAuthConfigurationInternal(context: Context, developerConfig: NativeAuthPublicClientApplicationConfiguration?): NativeAuthPublicClientApplicationConfiguration {
            val defaultMsalConfiguration: PublicClientApplicationConfiguration = PublicClientApplicationConfigurationFactory.loadConfiguration(context, R.raw.msal_default_config)
            // one authorities

            val defaultNativeAuthConfiguration = loadDefaultNativeAuthConfiguration(context)
           // no authorities

            val config = NativeAuthPublicClientApplicationConfiguration()
            config.appContext = context

            // Merge base MSAL defaults
            config.mergeConfiguration(defaultMsalConfiguration)  // authorities.size() == 1
            // Marge Native Auth defaults
            config.mergeConfiguration(defaultNativeAuthConfiguration)  // authorities.size() == 1

            config.authorities.clear()  // authorities.size() == 0 but authorities is not null

            // Check for developerConfig
            if (developerConfig != null) {  // skip the validation since developer config is null
                // Merge developer Native Auth configuration
                config.mergeConfiguration(developerConfig)
                // Validate Native Auth configuration
                config.validateConfiguration()
            }
```
Then structure the authority from the URL and append it into the authorities list
```kotlin
private static NativeAuthPublicClientApplication createNativeAuthApplication(@NonNull final NativeAuthPublicClientApplicationConfiguration config,
                                                                                 @Nullable final String clientId,
                                                                                 @Nullable final String authorityURL,
                                                                                 @Nullable final String redirectUri,
                                                                                 @Nullable final List<String> challengeTypes) throws BaseException {

        if (authorityURL != null) {
            config.getAuthorities().clear();

            final Authority authorityObject = Authority.getAuthorityFromAuthorityUrl(authorityURL, clientId);
            authorityObject.setDefault(true);
            config.getAuthorities().add(authorityObject);
        }
```

When passing in Config using ConfigFile and ResourceId, if `authoriries` field doesn't exist in the JSON file, the `gson.fromJson(config, NativeAuthPublicClientApplicationConfiguration::class.java)` will initialize it to null and fail the validation.

Yuki-YuXin replied:
Please refer to https://github.com/AzureAD/microsoft-authentication-library-for-android/pull/2313 for the new solution.

Yuki-YuXin commented:
Fixes [AB#{3287604}](https://dev.azure.com/IdentityDivision/Engineering/_workitems/edit/3287604)

Yuki-YuXin committed:
Authority and Command add capabilities

Yuki-YuXin commented on msal/src/main/java/com/microsoft/identity/nativeauth/statemachine/states/SignInStates.kt:
```diff
@@ -287,7 +287,17 @@ class SignInCodeRequiredState internal constructor(
                     }
 
-                    is INativeAuthCommandResult.Redirect, is INativeAuthCommandResult.APIError -> {
+                    is INativeAuthCommandResult.Redirect -> {
```
The repetition is to enumerate the possible scenarios under each method. Tend to keep the existing.

Yuki-YuXin commented on msal/src/main/java/com/microsoft/identity/nativeauth/NativeAuthPublicClientApplicationConfiguration.kt:
```diff
@@ -194,6 +218,28 @@ public class NativeAuthPublicClientApplicationConfiguration :
+    @Throws(MsalClientException::class)
+    private fun validateCapabilities() {
+        // Make all capabilities lowercase for simplicity
+        capabilities = capabilities?.map { it.lowercase() }
```
VALID_CAPABILITIES are defined lowercase constants.

Yuki-YuXin committed:
update latest feature branch common submodule reference

nilo-ms commented on msal/src/test/java/com/microsoft/identity/client/e2e/tests/network/nativeauth/SignInMFATest.kt:
```diff
@@ -331,4 +336,57 @@ class SignInMFATest : NativeAuthPublicClientApplicationAbstractTest() {
+
+                // SignIn specifying authentication context as claim
+                val result = application.signIn(params)
+                assertResult<SignInResult.MFARequired>(result)
```
the expected result here is "redirect" no "MFA_required"

Yuki-YuXin replied:
Done. And updated the others.

nilo-ms commented on msal/src/test/java/com/microsoft/identity/client/e2e/tests/network/nativeauth/SignInMFATest.kt:
```diff
@@ -331,4 +336,57 @@ class SignInMFATest : NativeAuthPublicClientApplicationAbstractTest() {
+     * - Return redirect with reason mfa required was not supplied
+     *
+     */
+    @Ignore("Retrieving OTP code failure.")
```
no OTP code should be sent for this test, so it should be enabled

Yuki-YuXin replied:
Done.

Yuki-YuXin replied:
Done. Use "Backward compatibility feature not available in eSTS production" instead.

nilo-ms commented on msal/src/test/java/com/microsoft/identity/nativeauth/NativeAuthPublicClientApplicationConfigurationTest.kt:
```diff
@@ -290,4 +290,97 @@ class NativeAuthPublicClientApplicationConfigurationTest {
+    }
+
+    @Test
+    fun testRepeatedCorrectCapabilities() {
```
what this test is exactly testing? Where are you checking that capabilities duplicates are removed?

Yuki-YuXin replied:
That's a good point. Add `Assert.assertEquals(spyConfig.getCapabilities()?.size,2)` and also for challenge types.

Yuki-YuXin committed:
Address comments: 1) update the comment 2) add redirect reason assertion

Yuki-YuXin committed:
Use parameter instead of config when initialization (#2313)

Context:
Deal with a comment under

https://github.com/AzureAD/microsoft-authentication-library-for-android/pull/2305

Yuki-YuXin committed:
update submodule reference

Yuki-YuXin committed:
Merge branch 'dev' into feature/backward-compatibility

# Conflicts:
#	common

Yuki-YuXin committed:
Merge branch 'dev' into feature/backward-compatibility

# Conflicts:
#	common

Yuki-YuXin committed:
Address comments

mohitc1 commented on msal/src/main/java/com/microsoft/identity/client/PublicClientApplicationConfiguration.java:
```diff
@@ -82,6 +82,7 @@
 
 import java.security.MessageDigest;
 import java.security.NoSuchAlgorithmException;
+import java.util.ArrayList;
```
Do you need this anymore?

Yuki-YuXin replied:
After checking. Nope. Removed this excess lines.

mohitc1 commented:
Don't see any issues with the three files you asked for review. Let me know if there's was specific part you were looking review for. 

Can you get this in ASAP?

fadidurah commented on msal/src/main/java/com/microsoft/identity/client/PublicClientApplication.java:
```diff
@@ -237,6 +237,7 @@ static class NONNULL_CONSTANTS {
         static final String AUTHORITY = "authority";
         static final String REDIRECT_URI = "redirect_uri";
         static final String CONFIG_FILE = "config_file";
+        static final String CLIENT_PARAMETER = "client_parameter";
```
Client_parameter? Should this be "capabilities"?

Yuki-YuXin replied:
In this PR, a new parameter class is introduced to all those config parameters (also capabilities) to provide extension for the future.

fadidurah reviewed (commented):
Reviewed MSAL-only code and LGTM, no review done for native auth code.

Yuki-YuXin committed:
Address comments

spetrescu84 commented on msal/src/test/java/com/microsoft/identity/client/e2e/tests/network/nativeauth/SignInJITTest.kt:
```diff
@@ -245,4 +247,49 @@ class SignInJITTest : NativeAuthPublicClientApplicationAbstractTest() {
+                // Return redirect with reason registration required was not supplied
+                assertTrue(signWithContinuationResult is SignInError)
+                assertTrue((signWithContinuationResult as SignInError).isBrowserRequired())
+                assertTrue(signWithContinuationResult.errorMessage!!.contains("registration required was not supplied"))
```
Is this the message that comes back from the server? The format looks strange.

Yuki-YuXin replied:
This is a placeholder when the test slice didn't work fine before. 
Done. Updated with "missing registration_required" which is more generic.

spetrescu84 replied:
So the test slice returns "missing registration_required" or we can't test? I see @nilo-ms [only checks that is not nil](https://github.com/AzureAD/microsoft-authentication-library-for-objc/pull/2648/files). Maybe do that also here? I am saying this that once we remove the ignore above this test will automatically start working. We can leave it to fix it then.

Yuki-YuXin replied:
![Screenshot 2025-06-23 112641](https://github.com/user-attachments/assets/f0304739-98d7-41fe-baab-6a658720e7fa)
Checking is not nil is a vague choice after the server response is specific.

spetrescu84 commented on msal/src/test/java/com/microsoft/identity/client/e2e/tests/network/nativeauth/SignInMFATest.kt:
```diff
@@ -331,4 +336,43 @@ class SignInMFATest : NativeAuthPublicClientApplicationAbstractTest() {
+                // Return redirect with reason mfa required was not supplied
+                assertTrue(result is SignInError)
+                assertTrue((result as SignInError).isBrowserRequired())
+                assertTrue(result.errorMessage!!.contains("mfa required was not supplied"))
```
Is this the message that comes back from the server? The format looks strange.

Yuki-YuXin replied:
This is a placeholder when the test slice didn't work fine before. 
Done. Updated with "missing registration_required" which is more generic.

spetrescu84 replied:
So the test slice returns "missing registration_required" or we can't test? I see @nilo-ms [only checks that is not nil](https://github.com/AzureAD/microsoft-authentication-library-for-objc/pull/2648/files). Maybe do that also here? I am saying this that once we remove the ignore above this test will automatically start working. We can leave it to fix it then.

spetrescu84 replied:
Also just in case, this one is about MFA.

Yuki-YuXin replied:
Yep. missing mfa_required

spetrescu84 commented on msal/src/test/java/com/microsoft/identity/nativeauth/NativeAuthPublicClientApplicationConfigurationTest.kt:
```diff
@@ -289,5 +290,100 @@ class NativeAuthPublicClientApplicationConfigurationTest {
+        whenever(spyConfig.defaultAuthority).thenReturn(NativeAuthCIAMAuthority(ciamAuthority, clientId))
+        whenever(spyConfig.isSharedDevice).thenReturn(false)
+        whenever(spyConfig.useBroker).thenReturn(false)
+        spyConfig.setCapabilities(listOf("lorem"))
```
Is this part of another PR where only specific capabilities work?

Yuki-YuXin replied:
Pardon. This is part of the sub-pr under https://github.com/AzureAD/microsoft-authentication-library-for-android/pull/2305/commits/6d553466cab96eb9642f5dd65e63d3492b95d8b3#diff-1e72e28f17fc1763d76a4b809b06adc0afb3fbf660cde3f7374282ab0547359c

Yuki-YuXin replied:
"lorem" is kind of invalid string practice I referred from other test cases.

spetrescu84 replied:
Found this check, all good - `if (capability !in VALID_CAPABILITIES) {`
